Dev python/spyder  Bug fix and version bump

AndrewAmmerlaan opened the pull request:
Bug: https://bugs.gentoo.org/702458

As of this commit: https://github.com/gentoo/gentoo/commit/3613dcb11fe5f8ed7d58f2df2fabb7609679e1b6

Python 2.7 support has been removed from spyder, this has caused the spyder ebuild to no longer install a executable 'spyder'. It used to install 'spyder2', 'spyder' and 'spyder3', where 'spyder' was linked to 'spyder2'.

Currently we have in the ebuild (spyder-4.0.0_beta1):
line 60: `make_desktop_entry spyder Spyder spyder "Development;IDE"`

Because the executable 'spyder' does not exist, this fails to create a working .desktop file. I therefore have changed this to: 
`make_desktop_entry spyder3 Spyder spyder "Development;IDE"`

I have also updated the spyder ebuild to 4.0.0 (and some other packages which `pip check` told me to update as well)

AndrewAmmerlaan commented:
> @gentoo/github: Too many disjoint maintainers, disabling auto-assignment.
Alright, I'll do it myself then :P

@gentoo/python
@aidecoe 
@gentoo/biology
@gentoo/science
@AndreyGrozin

flappyports commented on dev-python/sortedcontainers/sortedcontainers-2.1.0.ebuild:
```diff
@@ -0,0 +1,25 @@
+DEPEND="${RDEPEND}
+	test? ( dev-python/pytest[${PYTHON_USEDEP}] )"
+
+python_test() {
```
This can be dropped in favor of calling distutils_enable_tests pytest in global scope. See distutils-r1.eclass for an example

AndrewAmmerlaan replied:
Fixed in new pull request: https://github.com/gentoo/gentoo/pull/14014

flappyports commented on dev-python/sortedcontainers/sortedcontainers-2.1.0.ebuild:
```diff
@@ -0,0 +1,25 @@
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="test"
+RESTRICT="!test? ( test )"
```
Once you call distutils_enable_tests this can be dropped as it will be inherited from the eclass

flappyports commented on dev-python/sortedcontainers/sortedcontainers-2.1.0.ebuild:
```diff
@@ -0,0 +1,25 @@
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+PYTHON_COMPAT=( python2_7 python3_{5,6,7} )
```
py3_5 can be dropped as it will be dropped from Gentoo soon.

flappyports commented:
@AndrewAmmerlaan, overall, I see lots of good things being done here. Unfortunately, this is a *massive* PR with changes to multiple packages across the tree outside of just dev-python/spyder.

Further, there are trivial fixes as single commits... e.g. trailing line fixes etc.

Please break this up into multiple PRs touching single cat/pkg. This will allow the review to be much easier and allow the PR's to be merged. Only include multiple changes if the commits are required to be done at the same time.

AndrewAmmerlaan commented:
> @AndrewAmmerlaan, overall, I see lots of good things being done here. Unfortunately, this is a _massive_ PR with changes to multiple packages across the tree outside of just dev-python/spyder.
> 
> Further, there are trivial fixes as single commits... e.g. trailing line fixes etc. Please squash these.
> 
> Please break this up into multiple PRs touching single cat/pkg. This will allow the review to be much easier and allow the PR's to be merged. Only include multiple changes if the commits are required to be done at the same time.

@flappyports I have moved all packages to their own pull request, and squashed all commits into one single commit. Though I suspect the gentoo-repo-qa-bot will start complaining about missing dependencies now.

flappyports commented:
@AndrewAmmerlaan, I see the CI checks are still pending. If a pkg has dependencies that require other items in the same PR that is fine. This is why I said, " Only include multiple changes if the commits are required to be done at the same time."

AndrewAmmerlaan commented:
> @AndrewAmmerlaan, I see the CI checks are still pending. If a pkg has dependencies that require other items in the same PR that is fine. This is why I said, " Only include multiple changes if the commits are required to be done at the same time."

@flappyports That is why all those packages where in here to begin with.... they all are dependencies of spyder or dependencies of dependencies of spyder. I used `pip show` and `pip check` to obtain all the dependencies and upgraded/adjusted the necessary ebuilds accordingly until `pip check` no longer complained about incorrect/missing dependencies. I know before hand the CI checks will fail because `repoman ci` was complaining about the missing dependencies.

AndrewAmmerlaan commented:
> Because the executable 'spyder' does not exist, this fails to create a working .desktop file. I therefore have changed this to:
make_desktop_entry spyder3 Spyder spyder "Development;IDE"

So turns out the patch was causing the problem. It was patching out part of the code that was supposed to install `spyder` and the `spyder3.desktop` file. Combined with the fact that the ebuild made a `spyder.desktop` file pointing to non-existent `spyder` made the whole thing broken.

I cleaned up the patch and the ebuild. Now it installs the [scripts and .desktop file](https://github.com/spyder-ide/spyder/tree/master/scripts) provided by spyder.

I also added a version bump for h5py.

AndrewAmmerlaan commented:
@flappyports 

Bumped spyder to version 4.0.1

AndrewAmmerlaan commented:
@flappyports @juippis 

Sorry to @ you like this again, but could someone please take a look at this.

I put a lot of work into this, and it saddens me to see the PR collecting dust like this :(

The current spyder ebuilds in the tree are broken, the fix is right here!

juippis commented on dev-python/h5py/h5py-2.10.0.ebuild:
```diff
@@ -0,0 +1,64 @@
+inherit distutils-r1 flag-o-matic
+
+DESCRIPTION="Simple Python interface to HDF5 files"
+HOMEPAGE="http://www.h5py.org/"
```
`https://` whenever available. Also please include their Github page here.

juippis commented on dev-python/h5py/h5py-2.10.0.ebuild:
```diff
@@ -0,0 +1,64 @@
+	dev-python/pkgconfig[${PYTHON_USEDEP}]
+	doc? (
+		dev-python/alabaster[${PYTHON_USEDEP}]
+		>=dev-python/sphinx-1.3.1[${PYTHON_USEDEP}]
```
This is **most likely** a BDEPEND and you don't need PYTHON_USEDEP for it then. Also check how alabaster is needed (you can leave PYTHON_USEDEP for that).

EDIT: you can maybe use `distutils_enable_sphinx` here, but I don't have time to check this ebuild throughly and confirm it. I have a strong feeling it should work.

AndrewAmmerlaan replied:
`distutils_enable_sphinx` works but only with `--no-autodoc`

juippis commented on dev-python/h5py/h5py-2.10.0.ebuild:
```diff
@@ -0,0 +1,64 @@
+
+DEPEND="
+	${RDEPEND}
+	dev-python/cython[${PYTHON_USEDEP}]
```
This is an interesting dep. Could you verify how this program uses it? 
Calls cython during building -> BDEPEND without PYTHON_USEDEP.
Libs & headers needed during build-/runtime -> DEPEND/RDEPEND.

AndrewAmmerlaan replied:
Several files in h5py/h5py have references to cython, therefore I think it should DEPEND on cython. Pkgconfig on the other hand is only called in setup.......py files, so I have moved it to BDEPEND

juippis commented on dev-python/pytest-qt/metadata.xml:
```diff
@@ -0,0 +1,16 @@
+	<maintainer type="person">
+		<email>leycec@gmail.com</email>
+		<name>leycec</name>
+	</maintainer>
```
Maybe I missed something, but who is this person? Do you have their permission to be added here? We'd like them to ack this.

AndrewAmmerlaan replied:
@leycec of the [raiagent overlay](https://github.com/leycec/raiagent). The ebuild here is heavily based on the one found in the raiagent overlay, that's why they are still in the metadata file. I figured it would be impolite to just remove them from the metadata file.

@leycec do you want to be removed from pytest-qt's metadata file here, or would you like to keep it like this?

juippis replied:
When we push it to ::gentoo, anyone listed here will become the package maintainer and gets assigned bugs, etc is expected to keep answering to the bugs, bump requests and other ebuild maintenance routine.

Ie if you haven't talked to him about this at all, remove him from here. He can be added later with proper consent, or he can start making PRs and add himself.

juippis commented on dev-python/pytest-qt/pytest-qt-3.3.0.ebuild:
```diff
@@ -0,0 +1,38 @@
+	doc? (
+		dev-python/sphinx[${PYTHON_USEDEP}]
+		dev-python/sphinx_rtd_theme[${PYTHON_USEDEP}]
+	)"
```
You should be able to replace this with `distutils_enable_sphinx _build/html sphinx_rtd_theme`.

AndrewAmmerlaan replied:
Oops, I had changed this and pushed it to https://github.com/gentoo/gentoo/pull/14007 but I seem to have forgotten to push it to here as well.

juippis commented on dev-python/python-jsonrpc-server/python-jsonrpc-server-0.3.2.ebuild:
```diff
@@ -0,0 +1,29 @@
+
+IUSE="examples"
+
+DEPEND="dev-python/future[${PYTHON_USEDEP}]
```
With python packages, the correct is 99 % that this is a RDEPEND and not DEPEND ;)

Upstream enables python tests apparently, could you implement that in the ebuild?

juippis commented on dev-python/python-language-server/python-language-server-0.31.2.ebuild:
```diff
@@ -0,0 +1,33 @@
+DEPEND="dev-python/python-jsonrpc-server[${PYTHON_USEDEP}]
+		dev-python/jedi[${PYTHON_USEDEP}]
+		dev-python/pluggy[${PYTHON_USEDEP}]
+		dev-python/ujson[${PYTHON_USEDEP}]"
```
Please sort alphabetically and confirm these are not needed during runtime but during build-time.

I see upstream has some tests packaged, can you try enabling them?

juippis commented on dev-python/spyder-kernels/spyder-kernels-1.8.1.ebuild:
```diff
@@ -0,0 +1,24 @@
+		>=dev-python/ipykernel-5.1.3[${PYTHON_USEDEP}]
+		dev-python/pyzmq[${PYTHON_USEDEP}]
+		>=dev-python/jupyter_client-5.3.4[${PYTHON_USEDEP}]
+		dev-python/cloudpickle[${PYTHON_USEDEP}]"
```
Please sort alphabetically and check whether RDEPEND or DEPEND is correct.

https://github.com/spyder-ide/spyder-kernels/blob/master/ci/runtests.sh

juippis commented on dev-python/wurlitzer/wurlitzer-2.0.0.ebuild:
```diff
@@ -0,0 +1,16 @@
+inherit distutils-r1
+
+DESCRIPTION="Capture C-level stdout/stderr in Python "
+HOMEPAGE="https://github.com/minrk/wurlitzer"
```
https://github.com/minrk/wurlitzer/blob/master/test.py

Also add pypi page here.

AndrewAmmerlaan replied:
> https://github.com/minrk/wurlitzer/blob/master/test.py

I have added tests support, just `distutils_enable_tests pytest wasn't enough. I had to manually specify the `test.py` file for some reason.

juippis commented on dev-python/wurlitzer/wurlitzer-2.0.0.ebuild:
```diff
@@ -0,0 +1,16 @@
+
+inherit distutils-r1
+
+DESCRIPTION="Capture C-level stdout/stderr in Python "
```
Extra space in the end.

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -1,67 +1,64 @@
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI=6
+EAPI=7
```
Don't use `git mv` on this, but rather make one commit for a whole new file addition, then one new commit for removing 4.0.0_beta1.

AndrewAmmerlaan replied:
Fixed, should I remove the other old versions of spyder and spyder-kernels as well? Because some of those versions are quite old.

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -1,67 +1,64 @@
 
-inherit eutils distutils-r1 xdg-utils
-MY_P=${P/_beta/b}
+inherit eutils distutils-r1 xdg-utils desktop
```
Sort please.

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -1,67 +1,64 @@
-MY_P=${P/_beta/b}
+inherit eutils distutils-r1 xdg-utils desktop
 
 DESCRIPTION="The Scientific PYthon Development EnviRonment"
```
I checked how upstream writes their slogan, and looks like P**Y**thon ... Envi**R**onment are typoes.

AndrewAmmerlaan commented:
> Hey,
> 
> we hate to see work rot as well. But truth is, there is a lot of backlog and as @flappyports said, this is a massive PR requiring lots of time to go through. Especially for new packages we need to check upstream build systems, licenses, issues and such. I couldnt check spyder ebuild itself, because Github only shows changes when you do `git mv`.
> 
> I did my best to notice everything in one go, overall these look good.

Thank you for your review :)

I think I made all the changes you suggested, please let me know if I have missed anything.

Regarding `test` on spyder, spyder-kernels, python-language-server, python-jsonrpc-server. I have looked into enabling these some time ago. However, the tests in those packages require dev-python/pytest-cov, they do not work at all without it. Pytest-cov is a deprecated dependency and @flappyports told me not to add it to `DEPEND`, see `package.deprecated`:
> # Michał Górny <mgorny@gentoo.org> (2019-12-09)
> # Test coverage reporting is not useful for Gentoo ebuilds.  The deps
> # against coverage plugins are either entirely spurious, or the build
> # systems need to be fixed to work without them.
> dev-python/nose-cover3
> dev-python/nosexcover
> dev-python/pytest-cov

AndrewAmmerlaan commented on dev-python/wurlitzer/wurlitzer-2.0.0.ebuild:
```diff
@@ -0,0 +1,25 @@
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+
+IUSE="test"
+RESTRICT="!test? ( test )"
```
@juippis  Why is the qa-bot telling me:

>missing RESTRICT="!test? ( test )" with IUSE=test

when I have this right here?

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -0,0 +1,76 @@
+	dev-python/sphinx[${PYTHON_USEDEP}]
+	>=dev-python/spyder-kernels-1.8.1[${PYTHON_USEDEP}]
+	dev-python/watchdog[${PYTHON_USEDEP}]
+	hdf5? ( dev-python/h5py[${PYTHON_USEDEP}] )"
```
`optfeature` is succesfully used in this ebuild, why wasn't this moved there? Why are webkit and webengine USE flags when they seemingly have no function in this ebuild, nor is either of them **forced**. 
https://wiki.gentoo.org/wiki/Project:Quality_Assurance/Policies#USE-Controlled_Optional_RDEPENDs

AndrewAmmerlaan replied:
I did some digging into the source files, and you were correct. hdf5 is only required for one specific (internal) plugin, therefore I have moved it to optfeature. I also went ahead and added ebuilds for the other (external) plugins. (sorry I made this already big PR even bigger)

Regarding the tests on some of those packages, as with spyder and spyder-kernels etc, these tests are coverage tests (depend on pytest-cov) therefore I have not enabled them.

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -0,0 +1,76 @@
+	>=dev-python/qtconsole-4.5.0[${PYTHON_USEDEP}]
+	>=dev-python/QtPy-1.5.0[${PYTHON_USEDEP},svg,webengine?,webkit?]
+	>=sci-libs/Rtree-0.8.3[${PYTHON_USEDEP}]
+	dev-python/sphinx[${PYTHON_USEDEP}]
```
This bothers me as well. Could you carefully go through upstream requirements and that it mathces the list in this ebuild?

AndrewAmmerlaan replied:
I went through it again, and it seems there were indeed some dependencies that shouldn't have been there, possibly these were dropped in the 4.0.0 or 4.0.1 update. It now matches exactly the dependencies as specified in the setup.py file.

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -0,0 +1,76 @@
+	>=dev-python/qdarkstyle-2.7[${PYTHON_USEDEP}]
+	>=dev-python/qtawesome-0.5.7[${PYTHON_USEDEP}]
+	>=dev-python/qtconsole-4.6.0[${PYTHON_USEDEP}]
+	>=dev-python/qtconsole-4.5.0[${PYTHON_USEDEP}]
```
Duplicate.

juippis commented on dev-python/python-jsonrpc-server/python-jsonrpc-server-0.3.2.ebuild:
```diff
@@ -0,0 +1,31 @@
+
+inherit distutils-r1
+
+DESCRIPTION="A Python 2 and 3 asynchronous JSON RPC server "
```
Extra space.

AndrewAmmerlaan commented:
> Thanks for your patience, still few minor issues now that I finally had a proper look at spyder ebuild.

@juippis Thanks, I fixed up the spyder ebuild, removed unnecessary USE flags, added some `optfeatures`. 

I found that spyder has several additional plugins and have added ebuilds for those as well.
However, I'm running into a small problem with the qa-bot:

```
spyder-terminal¶
--
0.2.4 | NonsolvableDepsInDev | nonsolvable depset(rdepend) keyword(~amd64) dev profile (default/linux/amd64/17.0/x32) (1 total): solutions: [ net-libs/nodejs, sys-apps/yarn ]
```

I'm not sure what it is complaining about, `spyder-terminals` `BDEPEND`s on both nodejs and yarn. What does the qa bot want me to change?

Sorry for making this PR even bigger, but there were several ebuilds in the tree that were either broken or missing py3_7, and I had to change and fix those as well.

AndrewAmmerlaan commented:
@juippis 

I have updated some of the sub-PRs with the changes I made here, and added some other sub-PRs:
https://github.com/gentoo/gentoo/pull/14342 (coloredlogs)
https://github.com/gentoo/gentoo/pull/14343 (h5py)
https://github.com/gentoo/gentoo/pull/14007 (qtawesome)
https://github.com/gentoo/gentoo/pull/14006 (python-language-server)
https://github.com/gentoo/gentoo/pull/14012 (spyder-kernels)

Could you please review those first (most of them you already reviewed in this PR). Once those are merged I'll rebase this PR, that should about half the size of this PR, and make this a bit more manageable. Together we'll get through this, step by step :)

juippis commented:
Appreciate your effort here. I'd rather take time with this to put a polished final product in tree, instead of putting something unfinished that needs to be fixed little by little everyday. I'll wait for CI checks to finish and continue with the other 2 PRs later today.

AndrewAmmerlaan commented:
> Appreciate your effort here. I'd rather take time with this to put a polished final product in tree, instead of putting something unfinished that needs to be fixed little by little everyday. I'll wait for CI checks to finish and continue with the other 2 PRs later today.

Thank you, I have rebased to include the now-merged PR's. I also want to get this as perfect as possible :) In fact I started this whole adventure because the spyder ebuild was broken, which was very annoying. Though the whole project was a bit more work then I expected, I do want to do this as good as possible.

I've split of some more commits to sub-PRs, leaving only the spyder-* packages here:
https://github.com/gentoo/gentoo/pull/14345
https://github.com/gentoo/gentoo/pull/14344
https://github.com/gentoo/gentoo/pull/14346
These are 2 new packages, and 2 fixes to (very) broken ebuilds that are dependencies of packages here.

AndrewAmmerlaan commented:
@juippis I'm posting a quick status-update here to regain some overview and structure (a sort of to-do list I guess):

https://github.com/gentoo/gentoo/pull/14012 spyder-kernels **merged** and working
https://github.com/gentoo/gentoo/pull/14006 python-lang-serv **merged** but `DISTUTILS_USE_SETUPTOOLS=rdepend` should be removed

https://github.com/gentoo/gentoo/pull/14007 qtawesome/pytest-qt **open**, fully functional I think (`test` and `doc`)
https://github.com/gentoo/gentoo/pull/14342 coloredlogs and deps **open** fully functional I think ('test` and `doc`)
https://github.com/gentoo/gentoo/pull/14345 pweave/markdown **open** should work
https://github.com/gentoo/gentoo/pull/14344 line_profiler and deps **open** should work
https://github.com/gentoo/gentoo/pull/14343 h5py **open** works but `test` is broken
https://github.com/gentoo/gentoo/pull/14346 xterm-js **open** broken, I need to figure out how to correctly install a nodejs/npm application

AndrewAmmerlaan commented:
@juippis I used my newly-obtained knowledge on patching out pytest-cov dependencies to fix the tests on spyder and spyder-kernels, it's working for me now, does it work for you too?. It requires one extra dependencie: pytest-lazy-fixtures (https://github.com/gentoo/gentoo/pull/14479) 

I also added [an ebuild for the correct jedi](https://github.com/gentoo/gentoo/pull/14475) version required by spyder (0.14.1). The setup.py file explicitly asks for this version and has comment: '`This is here until Jedi 0.15+ fixes completions for Numpy and Pandas`'

juippis commented:
After jedi gets merged, what's still needed? Can you give me a list of open PRs here?

AndrewAmmerlaan commented:
> After jedi gets merged, what's still needed? Can you give me a list of open PRs here?

Here's the updated list:

https://github.com/gentoo/gentoo/pull/14012 spyder-kernels **merged**
https://github.com/gentoo/gentoo/pull/14006 python-lang-serv **merged** 
https://github.com/gentoo/gentoo/pull/14007 qtawesome/pytest-qt **merged**
https://github.com/gentoo/gentoo/pull/14342 coloredlogs and deps **merged**
https://github.com/gentoo/gentoo/pull/14345 pweave/markdown **merged**
https://github.com/gentoo/gentoo/pull/14343 h5py **merged**
https://github.com/gentoo/gentoo/pull/14479 pytest-lazy-fixture **merged**
https://github.com/gentoo/gentoo/pull/14475 jedi **merged**
~https://github.com/gentoo/gentoo/pull/14346 xterm-js **closed** broken, I need to figure out how to correctly install a nodejs/npm application~ [EDIT] Doesn't seem to be a requirement after all, spyder-terminal is very confusing about what it depends on :(
Required by spyder-memory-profiler and spyder-line-profiler:
https://github.com/gentoo/gentoo/pull/14344 line_profiler and deps **open** should work

Optionally not strictly required to merge spyder, version bump of python-language-server and python-jsonrpc-server, also contains [bug](https://bugs.gentoo.org/706420) fix:
https://github.com/gentoo/gentoo/pull/14463 **open**

AndrewAmmerlaan commented:
@juippis I just rebased this PR, there are only spyder-* ebuilds left here. It's still 12 commits, so it is still quite big, sorry about that :(

I went through the ebuilds here again last week, I enabled tests where possible. Some of the plugins do not have tests or I couldn't get them to work inside emerge/virtx.

Please, let me know if you find any issues

leycec commented:
@AndrewAmmerlaan: That was so thoughtful! Thanks for mentioning me and, more importantly, attempting to push these contributions into the Portage tree. `dev-python/pytest-qt` was recently removed due to ebuild rot and lack of developer interest; reverting that decision for the collective good of the Python community is awesome, indeed. You deserve all the gold stars! :star:  :star2: 

Sadly, I *really* lack the volunteer time to proxy-maintain `dev-python/pytest-qt`. As @juippis suggests, it'd be preferable to omit me from `metadata.xml` entirely – unless Gentoo *absolutely* cannot find anyone else to maintain this package. In that case, I'd be happy to lend a helping development hand. Thanks again!

AndrewAmmerlaan commented:
> @AndrewAmmerlaan: That was so thoughtful! Thanks for mentioning me and, more importantly, attempting to push these contributions into the Portage tree. `dev-python/pytest-qt` was recently removed due to ebuild rot and lack of developer interest; reverting that decision for the collective good of the Python community is awesome, indeed. You deserve all the gold stars! star star2
> 
> Sadly, I _really_ lack the volunteer time to proxy-maintain `dev-python/pytest-qt`. As @juippis suggests, it'd be preferable to omit me from `metadata.xml` entirely – unless Gentoo _absolutely_ cannot find anyone else to maintain this package. In that case, I'd be happy to lend a helping development hand. Thanks again!

Pytest-qt is already back in the gentoo repos :smile:, it got merged in one of my sub-PRs (I split this up to make it more manageable): https://packages.gentoo.org/packages/dev-python/pytest-qt (latest version is 3.3.0, sadly I was not able to also add py3_8 here because dependency QtPy is missing py3_8 at the moment)
I'll be proxy-maintaining it myself, along with some other dependencies of spyder. 


@juippis A new version of spyder-terminal was just released, It adds support for spyder4+ and py3_7, I updated the ebuild accordingly. 

The problem with the manifest of line_profiler that you noticed in https://github.com/gentoo/gentoo/pull/14344 got reported on the bug tracker: https://bugs.gentoo.org/708948 
My new PR: https://github.com/gentoo/gentoo/pull/14604 should fix this, as well as fix tests when the package is not already installed. Perhaps we should merge that first, otherwise everyone who will emerge spyder-line-profiler will run into that bug.

AndrewAmmerlaan commented:
@juippis 

@heroxbd just merged the fix for line_profiler https://github.com/gentoo/gentoo/pull/14604 (Thank you!). So we should be all-clear now to merge this. :smiley:

juippis commented on dev-python/spyder-autopep8/spyder-autopep8-9999.ebuild:
```diff
@@ -0,0 +1,20 @@
+
+DESCRIPTION="Plugin to run the autopep8 python linter from within the spyder editor"
+HOMEPAGE="https://github.com/spyder-ide/spyder-autopep8"
+EGIT_REPO_URI="https://github.com/spyder-ide/spyder-autopep8.git"
```
We won't add any "only" live packages to tree, so if this is needed you have to find an archived snapshot tarball and package that as `-0_pre<DateOfLatestCommit>`, ie `0_pre20200125`

juippis commented on dev-python/spyder-autopep8/spyder-autopep8-9999.ebuild:
```diff
@@ -0,0 +1,20 @@
+
+RDEPEND="
+	dev-python/autopep8[${PYTHON_USEDEP}]
+	<dev-python/spyder-4.0.0[${PYTHON_USEDEP}]"
```
... although I do wonder if it's worth it to add these at all? Ideally you'd add spyder-4.0.1, then stabilize it after ~30 days and drop every older version.

juippis commented on dev-python/spyder-docs/spyder-docs-9999.ebuild:
```diff
@@ -0,0 +1,24 @@
+
+DESCRIPTION="Documentation for Spyder"
+HOMEPAGE="https://github.com/spyder-ide/spyder-docs"
+EGIT_REPO_URI="https://github.com/spyder-ide/spyder-docs.git"
```
Same here.

juippis commented on dev-python/spyder-docs/spyder-docs-9999.ebuild:
```diff
@@ -0,0 +1,24 @@
+SLOT="0"
+KEYWORDS=
+
+BDEPEND="dev-python/sphinx"
```
(If you do end upd putting snapshot of this, remember the eclass function for this one)

AndrewAmmerlaan replied:
I can't use the distutils functions because `inherit distutils-r1` complains that there is no `setup.py`. How about instead I put the docs part into the spyder ebuild (I will push the changed spyder ebuild once I made the other changes)? Then I can use the `distutils_enable_sphinx` function, and then I don't need a seperate ebuild just for docs.

juippis replied:
Without reading other comments for now, you can use `DISTUTILS_USE_SETUPTOOLS=no` to bypass that error.

juippis commented on dev-python/spyder-docs/spyder-docs-9999.ebuild:
```diff
@@ -0,0 +1,24 @@
+
+src_install() {
+	dodoc -r doc/html
+}
```
... that should handle these as well...

juippis commented on dev-python/spyder-notebook/spyder-notebook-0.2.1.ebuild:
```diff
@@ -0,0 +1,33 @@
+
+inherit distutils-r1
+
+DESCRIPTION="Jupyter notebook integration with Spyder "
```
Extra space in the end.

juippis commented on dev-python/spyder-notebook/spyder-notebook-0.2.1.ebuild:
```diff
@@ -0,0 +1,33 @@
+# Tests do not work inside virtx/emerge for some reason, core dumped
+
+python_install_all() {
+	use doc && dodoc -r doc/*
```
Depending on the size of this, I'd say you can just install them without having a USE flag here. You can use a global DOCS array and then it gets installed with `distutils-r1_python_install_all` and you can remove `python_install_all()` function from the ebuild.

juippis commented on dev-python/spyder-notebook/spyder-notebook-0.2.1.ebuild:
```diff
@@ -0,0 +1,33 @@
+	dev-python/requests[${PYTHON_USEDEP}]
+	>=dev-python/spyder-4.0.0[${PYTHON_USEDEP}]"
+
+# Tests do not work inside virtx/emerge for some reason, core dumped
```
Define the phase, then add `RESTRICT="test"` with this comment on top of it. It's easier for others to test it when it's defined in the ebuild, but just restricted.

juippis commented on dev-python/spyder-reports/spyder-reports-0.1.1.ebuild:
```diff
@@ -0,0 +1,30 @@
+
+inherit distutils-r1
+
+DESCRIPTION="Spyder plugin to render Markdown reports using Pweave as a backend "
```
Extra space.

juippis commented on dev-python/spyder-reports/spyder-reports-0.1.1.ebuild:
```diff
@@ -0,0 +1,30 @@
+	use doc && dodoc -r doc/*
+
+	distutils-r1_python_install_all
+}
```
Same comments as in the previous ebuild.

juippis commented on dev-python/spyder-reports/spyder-reports-0.1.1.ebuild:
```diff
@@ -0,0 +1,30 @@
+
+RDEPEND="dev-python/matplotlib[${PYTHON_USEDEP}]
+	dev-python/pweave[${PYTHON_USEDEP}]
+	<dev-python/spyder-4.0.0[${PYTHON_USEDEP}]"
```
Although, once again I wonder if its worth to add this ebuild at all.

juippis commented on dev-python/spyder-vim/spyder-vim-9999.ebuild:
```diff
@@ -0,0 +1,37 @@
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS=
```
This should be removed.

juippis commented on dev-python/spyder-vim/spyder-vim-9999.ebuild:
```diff
@@ -0,0 +1,37 @@
+SLOT="0"
+KEYWORDS=
+
+RDEPEND="app-editors/vim
```
Does it really require vim just to emulate its keybinds? :)

juippis commented on dev-python/spyder-vim/spyder-vim-9999.ebuild:
```diff
@@ -0,0 +1,37 @@
+KEYWORDS=
+
+RDEPEND="app-editors/vim
+	<dev-python/spyder-4.0.0[${PYTHON_USEDEP}]"
```
Once again, worth it?

juippis commented on dev-python/spyder-vim/spyder-vim-9999.ebuild:
```diff
@@ -0,0 +1,37 @@
+	use doc && dodoc -r doc/*
+
+	distutils-r1_python_install_all
+}
```
Depending on size etc. I'd say it's worth to make into a USE flag once it becomes multiple megabytes large, like +5 MB, or has hundreds of small files.

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -0,0 +1,134 @@
+	rm spyder/plugins/editor/widgets/tests/test_goto.py || die
+
+	# skip online test
+	rm spyder/widgets/github/tests/test_github_backend.py
```
`|| die`

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -0,0 +1,134 @@
+	sed -i -e 's:test_go_to_line_shortcut:_&:' \
+		spyder/plugins/editor/widgets/tests/test_shortcuts.py || die
+	sed -i -e 's:test_delete_line_shortcut:_&:' \
+		spyder/plugins/editor/widgets/tests/test_shortcuts.py || die
```
Make this into a one sed call please. 

There are multiple separate seds, should these be made into patch/patches instead?

AndrewAmmerlaan replied:
I was using sed here, because patches tend to break on version bumps. I version bumped some other packages earlier, and the patches designed to skip specific tests all broke. Also it is a lot easier to write/change a sed line then it is to adjust a patch. So unless there is a good reason to use a patch, I'd prefer to use sed.

juippis replied:
That is the reason. These patches should be reported and fixed upstream as much as possible. Having these sed lines make it so comfortable that after couple of bumps, you wont remember they are there. Then in some future bump they might actually break something without the maintainer noticing.

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -0,0 +1,134 @@
+		optfeature "Jupyter notebook support" dev-python/spyder-notebook
+		optfeature "System terminal inside spyder" dev-python/spyder-terminal
+		optfeature "Markdown reports using Pweave" dev-python/spyder-reports
+		optfeature "Documentation for spyder" dev-python/spyder-docs
```
Some of these deps had `<spyder-4` requirement. Please go through them.

AndrewAmmerlaan commented:
> Ideally you'd add spyder-4.0.1, then stabilize it after ~30 days and drop every older version.

Alright, let's do that, I got rid of the plugins that are not compatible with spyder-4 and up. I commented out the relevant `optfeature` lines instead of removing them entirely to remind myself that these plugins exist. I'll keep the spyder-vim, spyder-autopep8  and spyder-reports ebuilds in my localrepo for now.


> What a monstrous PR it still is :) I think you can split these optional dependencies into smaller PRs to be looked at **after** spyder is merged. There are a lot of questionable stuff on them for now, explained below. You can always edit spyder adding new `optfeature`, but I do understand some people might miss them unti 4.0.2 is released. Just thinking it'd make this faster, but up yours, fine for me either way.

It's already a bit smaller now that I removed the plugins that do not work with >=spyder-4 :D
Only 8 commits left (one of which is just the removal of spyder-4.0.0_beta so that one doesn't really count :P ).

I also moved `spyder-docs` into the spyder ebuild. It seems to me preferable to not have a separate ebuild for `docs`, it also means I can use `distutils_enable_sphinx` :smile: 
I fetch a specific commit from spyder-docs in `SRC_URI` and move those files into the spyder source files in `python_prepare_all`. It does require me to update which commit of spyder-docs to fetch on spyder version bumps, but I suppose I'll just have to remember to do that :P
What do you think, `docs` in a separate ebuild or not?

juippis commented:
> I also moved `spyder-docs` into the spyder ebuild. It seems to me preferable to not have a separate ebuild for `docs`, it also means I can use `distutils_enable_sphinx` 😄
> I fetch a specific commit from spyder-docs in `SRC_URI` and move those files into the spyder source files in `python_prepare_all`. It does require me to update which commit of spyder-docs to fetch on spyder version bumps, but I suppose I'll just have to remember to do that :P
> What do you think, `docs` in a separate ebuild or not?

In theory it sounds fine, I'll check ebuilds later.

a17r commented on dev-python/spyder-notebook/spyder-notebook-0.2.1.ebuild:
```diff
@@ -0,0 +1,38 @@
+RDEPEND="dev-python/nbformat[${PYTHON_USEDEP}]
+	dev-python/notebook[${PYTHON_USEDEP}]
+	dev-python/psutil[${PYTHON_USEDEP}]
+	dev-python/QtPy[webkit,${PYTHON_USEDEP}]
```
new package blocking deprecated Qt5WebKit cleanup really makes me sad.

juippis replied:
Not trying to step on to your toes, but why is it deprecated? Upstream looks pretty active to me, 
https://github.com/qtwebkit/qtwebkit/branches/active

juippis replied:
cf36947cb83d55e687b318bb04fdff1c59ab737c would this work for spyder-notebook as well? If not, it looks like an optional dep and it could be left out from this update...

a17r replied:
> Not trying to step on to your toes, but why is it deprecated? Upstream looks pretty active to me,

QtWebKit is deprecated since February 2015. What we are looking at right now is a fork that has not gotten further than alpha status, with one year of inactivity in between. That's not to say we aren't glad someone is doing the valuable work keeping it on life support while upstreams are porting away.

AndrewAmmerlaan replied:
> new package blocking deprecated Qt5WebKit cleanup really makes me sad.

spyder-notebook, seems to work just fine without QtPy[webkit] and PyQt5[webkit]. Tests pass (though still not inside emerge, it dumps core), and a quick test inside spyder confirms the plugin loads correctly.

I think the reason I added QtPy[webkit] initially was because the file `spyder-notebook-0.2.1/spyder_notebook/utils/templates//weclome.html` contains lines with `-webkit-......`. But as I said it works fine without webkit support built into QtPy, so I'm not sure what those lines are about.

a17r replied:
that's a false positive then indeed, there's no QtWebKit involved in html files.

juippis commented on dev-python/spyder/spyder-4.0.1.ebuild:
```diff
@@ -0,0 +1,141 @@
+	>=dev-python/qdarkstyle-2.7[${PYTHON_USEDEP}]
+	>=dev-python/qtawesome-0.5.7[${PYTHON_USEDEP}]
+	>=dev-python/qtconsole-4.6.0[${PYTHON_USEDEP}]
+	>=dev-python/QtPy-1.5.0[${PYTHON_USEDEP},svg,webengine]
```
I got some errors in test_phase that was resolved with `dev-python/PyQt5[webkit]` 

Then I started running into
```
============================================ ERRORS ============================================
___ ERROR collecting spyder/plugins/variableexplorer/widgets/tests/test_collectioneditor.py ____
spyder/plugins/variableexplorer/widgets/tests/test_collectioneditor.py:26: in <module>
    import pandas
/usr/lib64/python3.6/site-packages/pandas/__init__.py:26: in <module>
    from pandas._libs import (hashtable as _hashtable,
/usr/lib64/python3.6/site-packages/pandas/_libs/__init__.py:4: in <module>
    from .tslibs import (
/usr/lib64/python3.6/site-packages/pandas/_libs/tslibs/__init__.py:7: in <module>
    from .period import Period, IncompatibleFrequency
stringsource:105: in init pandas._libs.tslibs.period
    ???
E   AttributeError: type object 'pandas._libs.tslibs.period.array' has no attribute '__reduce_cython__'
____ ERROR collecting spyder/plugins/variableexplorer/widgets/tests/test_dataframeeditor.py ____
spyder/plugins/variableexplorer/widgets/tests/test_dataframeeditor.py:25: in <module>
    from pandas import (DataFrame, date_range, read_csv, concat, Index, RangeIndex,
/usr/lib64/python3.6/site-packages/pandas/__init__.py:26: in <module>
    from pandas._libs import (hashtable as _hashtable,
/usr/lib64/python3.6/site-packages/pandas/_libs/__init__.py:4: in <module>
    from .tslibs import (
/usr/lib64/python3.6/site-packages/pandas/_libs/tslibs/__init__.py:7: in <module>
    from .period import Period, IncompatibleFrequency
stringsource:105: in init pandas._libs.tslibs.period
    ???
E   AttributeError: type object 'pandas._libs.tslibs.period.array' has no attribute '__reduce_cython__'
___ ERROR collecting spyder/plugins/variableexplorer/widgets/tests/test_namespacebrowser.py ____
spyder/plugins/variableexplorer/widgets/tests/test_namespacebrowser.py:22: in <module>
    from spyder.plugins.variableexplorer.widgets.tests.test_collectioneditor import data_table
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:656: in _load_unlocked
    ???
<frozen importlib._bootstrap>:626: in _load_backward_compatible
    ???
/usr/lib64/python3.6/site-packages/_pytest/assertion/rewrite.py:304: in load_module
    exec(co, mod.__dict__)
spyder/plugins/variableexplorer/widgets/tests/test_collectioneditor.py:26: in <module>
    import pandas
/usr/lib64/python3.6/site-packages/pandas/__init__.py:26: in <module>
    from pandas._libs import (hashtable as _hashtable,
/usr/lib64/python3.6/site-packages/pandas/_libs/__init__.py:4: in <module>
    from .tslibs import (
/usr/lib64/python3.6/site-packages/pandas/_libs/tslibs/__init__.py:7: in <module>
    from .period import Period, IncompatibleFrequency
stringsource:105: in init pandas._libs.tslibs.period
    ???
E   AttributeError: type object 'pandas._libs.tslibs.period.array' has no attribute '__reduce_cython__'
```

And no idea how to fix them. When emerged with `FEATURES="-test"` program seems to work, but might get some runtime errors? Could you verify if `PyQt5[webkit]` is really needed, or is it just some remnant of untouched test files?

Also why not link `/usr/bin/spyder -> /usr/bin/spyder3`?

AndrewAmmerlaan replied:
> When emerged with FEATURES="-test" program seems to work, but might get some runtime errors? Could you verify if PyQt5[webkit] is really needed, or is it just some remnant of untouched test files?

It looks like QtPy[webkit]/PyQt5[webkit] is only required by the tests. I just did some work in spyder just now with webkit support removed from QtPy/PyQt5, and it worked just fine.
What to do now though? Should I add `DEPEND="test? ( QtPy[webkit] ) `, because apparently webkit is supposed to get removed from the tree? Or should I just remove the tests that try to import QtWebKit?

juippis replied:
Latter, if possible. I believe a version bump can be merged with webkit support, but would rather get rid of it sooner than later.

AndrewAmmerlaan replied:
I've added some lines to the ebuild to remove the problematic tests. I have tried a patch, however, repoman complained that the file size of my patch was to big, so I gave up on that and just added some lines to the python_prepare_all phase

juippis replied:
I could host the patch if you wish to, however looks like you already pushed your changes.

However you prefer.

AndrewAmmerlaan replied:
> I could host the patch if you wish to, however looks like you already pushed your changes.
>
> However you prefer.

Let's just do it like this for now, there already is quite a lot of lines in python_prepare_all and I'd prefer to keep it mostly in one place. That way I can just comment out the lines in the ebuild one by one when there is a version bump to test if those tests are still broken.

juippis commented on dev-python/spyder-kernels/spyder-kernels-1.8.1-r1.ebuild:
```diff
@@ -0,0 +1,45 @@
+			spyder_kernels/console/tests/test_console_kernel.py || die
+	fi
+
+	pytest -vv || die "Tests fail with ${EPYTHON}"
```
You don't really have to revbump to enable tests, you can just edit the old ebuild in place. That being said, I also get that weird pandas problem with this package, 
```
============================================ ERRORS ============================================
__________________ ERROR collecting spyder_kernels/utils/tests/test_nsview.py __________________
spyder_kernels/utils/tests/test_nsview.py:18: in <module>
    import pandas as pd
/usr/lib64/python3.6/site-packages/pandas/__init__.py:26: in <module>
    from pandas._libs import (hashtable as _hashtable,
/usr/lib64/python3.6/site-packages/pandas/_libs/__init__.py:4: in <module>
    from .tslibs import (
/usr/lib64/python3.6/site-packages/pandas/_libs/tslibs/__init__.py:4: in <module>
    from .conversion import normalize_date, localize_pydatetime, tz_convert_single
pandas/_libs/tslibs/conversion.pyx:234: in init pandas._libs.tslibs.conversion
    ???
E   AttributeError: type object 'pandas._libs.tslibs.conversion._TSObject' has no attribute '__reduce_cython__'
!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

Everything else works.

AndrewAmmerlaan replied:
>  That being said, I also get that weird pandas problem with this package

This is interesting, I don't get any errors about pandas in spyder or spyder-kernels. This is the pandas I have installed:
```
[I] dev-python/pandas
Available versions:  0.24.2^t **9999*l^t {X doc full-support minimal test PYTHON_TARGETS="python3_6 python3_7"}
Installed versions:  0.24.2^t(19:57:38 29/01/20)(X -doc -full-support -minimal -test PYTHON_TARGETS="python3_6 python3_7")
Homepage:            https://pandas.pydata.org/
Description:         Powerful data structures for data analysis and statistics
```

What flags do you have enabled on pandas? I have this suspicion that maybe USE="X" is required on pandas for this to work.

juippis replied:
I tried with `X`, `-X`, `full-support`, `-full-support` without changes. It also looks like cython has something to do with it. 

I tried googling the error, but so far none of the workarounds worked for me. Well, they were mostly just "update to latest version"

AndrewAmmerlaan replied:
Hmm, strange, I just re-emerged pandas, and now tests in spyder stopped working for me. I get the same error you're getting.

AndrewAmmerlaan replied:
Update: `MAKEOPTS="-j1" emerge -1 pandas` fixes this for me? Does that work for you to?

juippis replied:
Weirdly... yes.

AndrewAmmerlaan replied:
> Weirdly... yes.

Packages like scipy, and pandas have always been kinda unstable when it comes to parallel building. I remember that when I installed pandas I had to use `MAKEOPTS="-j1"`to get it to install at all, same for scipy (https://bugs.gentoo.org/676640, https://bugs.gentoo.org/680036). I often have to do the same when updating these packages. 

In fact, I was extremely surprised today to find that pandas compiled *without* `MAKEOPTS="-j1"`, when I found that spyder's test broke, I was immediately suspicious about `MAKEOPTS="-j1"`. Anyway, this is an issue that should probably be fixed in the pandas ebuild, so I have reported it here: https://bugs.gentoo.org/710030

juippis replied:
Damn you're efficient. I also made a bug about it, 
https://bugs.gentoo.org/710062

Well they should definitely notice it now! :)

juippis commented:
EDIT: I mean, attempt to use `DISTUTILS_SINGLE_IMPL` in future.

AndrewAmmerlaan commented:
> I made few changes, please check
> [2ce6f1b](https://github.com/gentoo/gentoo/commit/2ce6f1b741cba5f6bed8ca9b12ee07163b717487)
> [01b9714](https://github.com/gentoo/gentoo/commit/01b9714de85c236653df0e0da921f4ab6dc87c6a)
> 
> I tested them and should work.

Just tested it as well, works for me too!

> EDIT: I mean, attempt to use DISTUTILS_SINGLE_IMPL in future.

That is a good idea, there indeed doesn't seem to be any need to have spyder built against multiple python versions.

Thank you so much for your help and advice, I really appreciate it :)
